### PR TITLE
Fix bug in Balance stage when a downstream wait a response indefinitely

### LINKED ISSUE #25387: FixedSizeBuffer isEmpty return false when buffer is empty
The method `isEmpty` in class `FixedSizeBuffer` return true only when `used == 0`. But the method `used` can return a negative value when you `dequeue` more than `enqueue`.
This break sometime the Balance stage in the OutHandler logic for example.

### PR DESCRIPTION
**Pull request by guiCarlier**
Fix bug in Balance stage when some downstreams pull and close immediately and an another try to pull but wait a response indefinitely.
Resolve #25387

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @guiCarlier,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by johanandren** [comment]
OK TO TEST

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBalanceSpec.scala [hunk 1]
```diff
@@ -321,6 +321,44 @@ class GraphBalanceSpec extends StreamSpec {
...
+      p1.expectRequest(bsub, 16)
+      bsub.sendNext(1)
+
+      sub3.request(1)
+      c3.expectNext(1)
+
+      sub3.cancel()
+
+      bsub.expectCancellation()
+    }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBalanceSpec.scala]: Was this consistently failing without the fix or more racey-sometimes-failing?

Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBalanceSpec.scala]: Great, just curious.

**Review by johanandren — APPROVED** [review]
LGTM, thanks @guiCarlier !

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 2]
```diff
@@ -883,8 +883,8 @@ final class Balance[T](val outputPorts: Int, val waitForAllDownstreams: Boolean,
         if (!isClosed(out)) {
           push(out, grab(in))
           if (!noPending) pull(in)
-        } else {
-          // try to find one output that isn't closed
+        } else if (!noPending) {
+          // if they are pending outlets, try to find one output that isn't closed
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: ok, I was a bit surprised that FixedSizeBuffer doesn't guard agains invalid dequeue, but it's documented so I guess there is a good reason

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**